File: java/system_design/ObserverPattern.java
import java.util.*;

interface Observer { void update(Button b); }

class Button {
    List<Observer> observers = new ArrayList<>();
    boolean pressed;
    void attach(Observer o) { observers.add(o); }
    void detach(Observer o) { observers.remove(o); }
    void notify_obs() { for(Observer o : observers) o.update(this); }
    void press() { pressed=true; notify_obs(); }
    void release() { pressed=false; notify_obs(); }
}

class LogObserver implements Observer {
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public void update(Button b) { System.out.println("Button: " + (b.pressed ? "pressed" : "released")); }
}

/**
 * ObserverPattern - [Brief description]
 *
 * <p>OVERVIEW:
 * [Detailed explanation of what this class does]
 *
 * <p>COMPLEXITY:
 * <ul>
 *   <li>Time: [See method documentation]</li>
 *   <li>Space: O(n) where n is [the element count]</li>
 * </ul>
 *
 * <p>USAGE:
 * [How to use this class, with example]
 *
 * @author Interview Preparation
 * @since 1.0
 */

public class ObserverPattern {
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static void main(String[] args) {
        Button btn = new Button();
        btn.attach(new LogObserver());
        btn.press();
        btn.release();
    }
}

File: python/system_design/observer_pattern.py
"""Observer Pattern - Event publishing and subscription"""

class Observer:
    """Observer interface"""

        """update implementation.

        Time: O(n)
        Space: O(1)
        """
    def update(self, subject):
        raise NotImplementedError


class Subject:
    """Observable subject"""

    def __init__(self):
        self._observers = []

    def attach(self, observer: Observer):
        """Attach observer"""
        if observer not in self._observers:
            self._observers.append(observer)

    def detach(self, observer: Observer):
        """Detach observer"""
        if observer in self._observers:
            self._observers.remove(observer)

    def notify(self):
        """Notify all observers"""
        for observer in self._observers:
            observer.update(self)


class Button(Subject):
    """Concrete subject - Button"""

    def __init__(self):
        super().__init__()
        self._is_pressed = False

    def press(self):
        """Button pressed"""
        self._is_pressed = True
        self.notify()

    def release(self):
        """Button released"""
        self._is_pressed = False
        self.notify()

    @property
        """is_pressed implementation.

        Time: O(n)
        Space: O(1)
        """
    def is_pressed(self):
        return self._is_pressed


class LogObserver(Observer):
    """Concrete observer - Log events"""

    def __init__(self, name: str):
        self.name = name

        """update implementation.

        Time: O(n)
        Space: O(1)
        """
    def update(self, subject: Button):
        print(f"[{self.name}] Button is {'pressed' if subject.is_pressed else 'released'}")


class DisplayObserver(Observer):
    """Concrete observer - Display status"""

    def __init__(self):
        self.status = "released"

        """update implementation.

        Time: O(n)
        Space: O(1)
        """
    def update(self, subject: Button):
        self.status = "pressed" if subject.is_pressed else "released"
        print(f"[Display] Current status: {self.status}")


if __name__ == "__main__":
    button = Button()

    log_obs = LogObserver("Logger")
    display_obs = DisplayObserver()

    button.attach(log_obs)
    button.attach(display_obs)

    print("=== Pressing button ===")
    button.press()

    print("\n=== Releasing button ===")
    button.release()

    print("\n=== Detaching logger ===")
    button.detach(log_obs)
    button.press()